pkg/proxy/service_test.go

```diff
@@ -0,0 +1,534 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	apiservice "k8s.io/kubernetes/pkg/api/service"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const testHostname = "test-hostname"
+
+// fake implementation for service info.
+type fakeServiceInfo struct {
+	clusterIP           net.IP
+	port                int
+	protocol            api.Protocol
+	healthCheckNodePort int
+}
+
+func (f *fakeServiceInfo) String() string {
+	return fmt.Sprintf("%s:%d/%s", f.clusterIP, f.port, f.protocol)
+}
+
+func (f *fakeServiceInfo) ClusterIP() string {
+	return f.clusterIP.String()
+}
+
+func (f *fakeServiceInfo) Protocol() api.Protocol {
+	return f.protocol
+}
+
+func (f *fakeServiceInfo) HealthCheckNodePort() int {
+	return f.healthCheckNodePort
+}
+
+func makeTestServiceInfo(clusterIP string, port int, protocol string, healthcheckNodePort int) *fakeServiceInfo {
+	info := &fakeServiceInfo{
+		clusterIP: net.ParseIP(clusterIP),
+		port:      port,
+		protocol:  api.Protocol(protocol),
+	}
+	if healthcheckNodePort != 0 {
+		info.healthCheckNodePort = healthcheckNodePort
+	}
+	return info
+}
+
+func newFakeServiceInfo(svcPortName ServicePortName, servicePort *api.ServicePort, service *api.Service) ServicePort {
+	info := &fakeServiceInfo{
+		clusterIP: net.ParseIP(service.Spec.ClusterIP),
+		port:      int(servicePort.Port),
+		protocol:  servicePort.Protocol,
+	}
+	if apiservice.NeedsHealthCheck(service) {
+		p := service.Spec.HealthCheckNodePort
+		if p != 0 {
+			info.healthCheckNodePort = int(p)
+		}
+	}
+	return info
+}
+
+func makeTestService(namespace, name string, svcFunc func(*api.Service)) *api.Service {
+	svc := &api.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+		},
+		Spec:   api.ServiceSpec{},
+		Status: api.ServiceStatus{},
+	}
+	svcFunc(svc)
+	return svc
+}
+
+func addTestPort(array []api.ServicePort, name string, protocol api.Protocol, port, nodeport int32, targetPort int) []api.ServicePort {
+	svcPort := api.ServicePort{
+		Name:       name,
+		Protocol:   protocol,
+		Port:       port,
+		NodePort:   nodeport,
+		TargetPort: intstr.FromInt(targetPort),
+	}
+	return append(array, svcPort)
+}
+
+func makeNSN(namespace, name string) types.NamespacedName {
+	return types.NamespacedName{Namespace: namespace, Name: name}
+}
+
+func makeServicePortName(ns, name, port string) ServicePortName {
+	return ServicePortName{
+		NamespacedName: makeNSN(ns, name),
+		Port:           port,
+	}
+}
+
+func Test_serviceToServiceMap(t *testing.T) {
+	testCases := []struct {
+		service  *api.Service
+		expected map[ServicePortName]*fakeServiceInfo
+	}{
+		{
+			// Case[0]: nothing
+			service:  nil,
+			expected: map[ServicePortName]*fakeServiceInfo{},
+		},
+		{
+			// Case[1]: headless service
+			service: makeTestService("somewhere-else", "headless", func(svc *api.Service) {
```

Comment:
Instead of using "somewhere-else" for the namespace, can you use something a little clearer like "test-ns" or "test-namespace". Just something that indicates that the argument is a namespace so it is easier to read. You can make this change in quite a few places.